### ISSUE TITLE
Turn browser_params and manager_params into dataclasses and validate them

### ISSUE THREAD

**Issue by englehardt** [mozilla/OpenWPM#495]
A mistyped or unknown configuration parameter in `browser_params` or `manager_params` will currently not throw an error. This may lead to missing data that's not discovered until the end of a crawl. We should validate that the configuration parameters are valid before starting a measurement.

We can do this basic check by using [dataclasses](https://docs.python.org/3/library/dataclasses.html) and [dataclass-json](https://github.com/lidatong/dataclasses-json). This allows us to work with objects that have fields instead of having dicts.

These classes should have comments for each field explaining its use and type annotations to ensure proper usage.
We could then remove [`default_browser_params.json`](https://github.com/mozilla/OpenWPM/blob/7d118c9528d7bc94c44dd34f211802033b5ca9bd/automation/default_browser_params.json) and [`default_manager_params.json`](https://github.com/mozilla/OpenWPM/blob/7d118c9528d7bc94c44dd34f211802033b5ca9bd/automation/default_manager_params.json) and instead set the values there as default values in the class.

Once we have these classes we should then write three functions:
1. `validate_manager_params` which would accept a `ManagerParams` object
2. `validate_browser_params` which would accept a `BrowserParams` object
3. `validate_run_config` which accepts one `ManagerParams` and a list of `BrowserParams`

All of these should raise a `ConfigError` if anything is invalid.

We currently have a bunch of code spread throughout the code base that checks the validity of different configs at different points in time.

Examples:
-  [`save_content`](https://github.com/mozilla/OpenWPM/blob/7d118c9528d7bc94c44dd34f211802033b5ca9bd/automation/DeployBrowsers/deploy_firefox.py#L125-L129) - This belongs into `validate_run_config`
- [`browser`](https://github.com/mozilla/OpenWPM/blob/65337e0c19dc857f63bbca1b3aa3773a2c49fd42/automation/DeployBrowsers/deploy_browser.py#L7) - This belongs into `validate_browser_params`
- [`callstack_instrument`](https://github.com/mozilla/OpenWPM/blob/65337e0c19dc857f63bbca1b3aa3773a2c49fd42/automation/DeployBrowsers/deploy_firefox.py#L107-L113) - This belongs into `validate_browser_params`
- [`save_content`](https://github.com/mozilla/OpenWPM/blob/65337e0c19dc857f63bbca1b3aa3773a2c49fd42/automation/DeployBrowsers/deploy_firefox.py#L115-L125) - This belongs into `validate_browser_params`

More cases can be found by looking for where a `BrowserConfigError` gets raised in the code  base.

See #493 and #494 as examples of issues that arise from not having config validation.

**Comment by vringar** [comment]
#783 also could have been avoided if we validated the dict to only contain valid keys.

**Comment by vringar** [comment]
#584 should be addressed as part of this issue

**Comment by ankushduacodes** [comment]
@vringar Is it okay if I work on this?

**Comment by vringar** [comment]
Absolutely :)

**Comment by ankushduacodes** [comment]
@vringar I have not worked with dataclasses before, If I need any help I will let you know

**Comment by ankushduacodes** [comment]
@vringar I am gonna start working on this one now.
I just need to know few things:
1. Do you want me to combine both dataclasses into one file or seperate files?
2. Which directory you want me to keep the file/files in?
3. Are there any browser or manager params other than default one that I should be aware of?

**Comment by vringar** [comment]
1. I'd put them in one file and call that `config.py`
2. The file should be in the `openwpm` directory
3. No, I don't think there are any you need to be aware of

**Comment by ankushduacodes** [comment]
@vringar As mentioned in the issue description, that we should use [dataclasss-json](https://github.com/lidatong/dataclasses-json) but I am not sure why we need to use this library, Could you please explain what I am missing here?

**Comment by ankushduacodes** [comment]
@vringar  1. We do want to set the default values for browser and manager params right?

if so, Then there is an issue with setting default value of `js_instrument_settings`.
As you can see here:
https://github.com/mozilla/OpenWPM/blob/d74dcf75d777a5487676c496abd0baa0a0c76ea3/openwpm/default_browser_params.json#L5
default value for `js_instrument_settings` is a list(which is of mutable type), now In dataclasses (even in simple classes), we  can not set mutable default values as explained in the [dataclasses documentation](https://docs.python.org/3/library/dataclasses.html#mutable-default-values)

Now one way to do this is explained [here in the docs using dataclasses.field()](https://docs.python.org/3/library/dataclasses.html#dataclasses.field) but in this case as well we can only initialize the `js_instrument_settings` with an empty list.
In this case we can create a method something like:
```py
def add_default_js_instrument_setting(self):
    js_instrument_settings.append("collection_fingerprinting")
```
but this way we will have to call this `add_default_js_instrument_setting()` everytime if we need collection fingerprinting(I am not sure what this collection fingerprinting is for).


2. Are there any other options that can be added to the `js_instrument_settings` list? and where is this option required?

**Comment by vringar** [comment]
> @vringar As mentioned in the issue description, that we should use [dataclasss-json](https://github.com/lidatong/dataclasses-json) but I am not sure why we need to use this library, Could you please explain what I am missing here?

We currently allow for people to just provide their own JSON and use that instead of the `default_manager_params.json`, so that functionality should remain.

**Comment by vringar** [comment]
> 1. We do want to set the default values for browser and manager params right?

Yes.
>  if so, Then there is an issue with setting the default value of `js_instrument_settings`.
>    As you can see here:
>    https://github.com/mozilla/OpenWPM/blob/d74dcf75d777a5487676c496abd0baa0a0c76ea3/openwpm/default_browser_params.json#L5    
>   The default value for `js_instrument_settings` is a list (which is a mutable type), now in dataclasses (even in simple classes), we  can not set mutable default values as explained in the [dataclasses documentation](https://docs.python.org/3/library/dataclasses.html#mutable-default-values)

Thanks for finding this issue and doing the research on it.
Can you try doing something like:
```python
field(default_factory=lambda: ["collection_fingerprinting"] )
```
> 2. Do you have a list of params whose value can be other than the defaults with the possible value of each? (other than bools), It will help when implementing validations
>  for example:
>  browser can either be equal to "firefox" or "chrome"

Some of those can be found [here](https://github.com/mozilla/OpenWPM/blob/master/docs/Configuration.md#browser-configuration-options). If you find any field insufficiently documented please ask and then add the response to the documentation.
>  3. The values that are set to null are of what type? str?:

They represent paths, so we could try and see if we can use [`Path`](https://docs.python.org/3/library/pathlib.html#pathlib.Path)  but I suspect that won't work well with serialization and deserialization.
So yes, a str might be the best we can do here.

**Comment by ankushduacodes** [comment]
> ```python
> field(default_factory=lambda: ["collection_fingerprinting"] )
> ```
This works!... Thank you

**Comment by ankushduacodes** [comment]
> 3. `validate_run_config` which accepts one `ManagerParams` and a list of `BrowserParams`
@vringar Could you please explain what does @englehardt means by this line from the issue description?

**Comment by ankushduacodes** [comment]
@vringar I wondering if we could discuss the flow of 
1. When and where dataclasses will be instantiated
2. When and where are validations are being applied
3. What happens to the dataclasses if user provides its own browser and manager params
4. Where are do we need to convert dataclasses into its json counterparts
5. I see that currently default browser and manager params are being loaded with the help of `TaskManager.load_default_params()`, How does its functionality changes?
6. And of course this [comment](https://github.com/mozilla/OpenWPM/issues/495#issuecomment-732231312)

**Comment by vringar** [comment]
> 1. `validate_run_config` which accepts one `ManagerParams` and a list of `BrowserParams`
> @vringar Could you please explain what does @englehardt means by this line from the issue description?

I edited his comment and this line is from me.
I thought of something like
```python
from typing import List
def validate_run_config(manager_params: ManagerParams, browser_params: List[BrowserParams]):
    pass
```

**Comment by ankushduacodes** [comment]
> > 1. `validate_run_config` which accepts one `ManagerParams` and a list of `BrowserParams`
> >    @vringar Could you please explain what does @englehardt means by this line from the issue description?
> 
> I edited his comment and this line is from me.
> I thought of something like
> 
> ```python
> from typing import List
> def validate_run_config(manager_params: ManagerParams, browser_params: List[BrowserParams]):
>     pass
> ```

But what are run configs and why do we need them?

**Comment by vringar** [comment]
1. When and where dataclasses will be instantiated?

Somewhere before the TaskManager gets created. A piece of sample code I'm thinking of:
```python
manager_params = ManagerParams() # the default ManagerParams
browser_params = []
for _ in range(NUM_BROWSERS):
    bp = BrowserParams(display_mode = "headless")
    browser_params.append(bp)

task_manager = TaskManager(manager_params, browser_params)
```
 
2. When and where are validations are being applied

In the `__init__` of the `TaskManager`
Sample Code:
```python
def __init__(manager_params: ManagerParams, browser_params: List[BrowserParams]):
     validate_manager_params(manager_params)
     for bp in browser_params:
          validate_browser_params(bp)
    validate_run_config(manager_params, browser_params)
```
 3. What happens to the dataclasses if user provides its own browser and manager params

The user has to pass the browser and manager params in a way that they can pass our validation.
The suggested way to do that is to instantiate the dataclasses and change the values that you want to change.

4. Where are do we need to convert dataclasses into its json counterparts

I was mostly thinking about the other way around. If someone still has their custom `custom_browser_params.json`
I want them to be able to write:
```python
with open('custom_browser_params.json', 'r') as file:
    bp = BrowserParams.from_json(file.read())
```

5. I see that currently default browser and manager params are being loaded with the help of `TaskManager.load_default_params()`, How does its functionality changes?

I think these methods and the associated JSON files can be deleted.

> But what are run configs and why do we need them?

Run config is probably a bad word (I'm open for suggestions here). For me "run config" means
how an entire crawl is configured. What `validate_run_config` is supposed to check is e.g.:
- If you have content_saving enabled in one of the browser_params you should also have content_saving enabled
  in the manager_params (This is currently checked internally but shouldn't be)


As for validations I think there are some general expectations:
1. We don't aim to prevent all user mistakes from happening but instead give reasonable guidance
2. We won't implement a typechecking algorithm. If the user uses the wrong type the plattform will just error
eventually and the user can look at the docs or the dataclasses to see what type they should have used.

For strings I also wouldn't assert that they are strings but instead check that they are in the list of valid values.

**Comment by ankushduacodes** [comment]
Got it!... Could also please check it #807 , It has issues with new added dependencies which is making all the tests to fail

**Comment by vringar** [comment]
I'm looking into it!

**Comment by ankushduacodes** [comment]
@vringar I have added validations for browser params, please have look so I know I am on the right path

**Comment by ankushduacodes** [comment]
@vringar could you please help me with `js_instrument`, I tried reading the documentation but could not understand much.
I also not sure what kind validation will be applicable to `js_instrument_settings`.
Your guidance will be much appreciated.

**Comment by vringar** [comment]
I think the validation for the `js_instrument_settings` happens here:
https://github.com/mozilla/OpenWPM/blob/051a3846cbd62a50a462fa37333ab5eeae037b39/openwpm/task_manager.py#L128-L132
So you don't need to add validation for them.

**Comment by ankushduacodes** [comment]
Could you please give a gist of what js_instrument is for?

**Comment by vringar** [comment]
Browsers expose a [WebAPI](https://developer.mozilla.org/en-US/docs/Web/API) that websites can use to interact with the browser that displays the website.
OpenWPM enables user to see what WebAPIs are called in what way with the JS instrument, by allowing them to say which WebAPIs they are interested in and then logging every time a method is called or a property is accessed.

**Comment by ankushduacodes** [comment]
@vringar For the manager params, theres not much to validate in them, may be I can set validation for `log_file` extension type and `database_name` extension type, other than that there's directories for data and logs to be stored, Do we want to validate these directories to see if they actually are directories?
What else is there to validate in manager params?

**Comment by vringar** [comment]
What you suggest seems reasonable. There isn't much to validate in there and there will be even less once #753 lands so don't invest too much energy there.

**Comment by ankushduacodes** [comment]
@vringar I have tried something to validate file extension which tbh looks kinda ugly, what is your opinion on the latest commit. I am not a fan of what I have done there but I could not think of something else

**Comment by vringar** [comment]
I much prefer the splitext to a regex and I don't think it looks ugly.
The code expresses what we want to do and is easily readable.

**Comment by ankushduacodes** [comment]
@vringar After I delete the following check the file will be redundent which will just be calling one method, 
https://github.com/mozilla/OpenWPM/blob/65337e0c19dc857f63bbca1b3aa3773a2c49fd42/automation/DeployBrowsers/deploy_browser.py#L5-L15
`deploy_browser()` is only being used in one place here: https://github.com/mozilla/OpenWPM/blob/051a3846cbd62a50a462fa37333ab5eeae037b39/openwpm/browser_manager.py#L436-L438

Should I just refactor the code for it `browser_manager.py` to call `deploy_firefox` directly?

**Comment by vringar** [comment]
You can delete `deploy_browser.py` and just call `deploy_firefox` in `browser_manager.py`.
OpenWPM won't support any other browser in the foreseeable future

**Comment by vringar** [comment]
Please also delete `screen_resolutions.txt` and `ua_strings.txt` while you are at it.

**Comment by ankushduacodes** [comment]
One more important question for the night and then I am done 😅, 
This issue was filed because `save_content = ['script']` didn’t raise any issue as described in #494.
This is still the case with the logic below:
```py
    if browser_params.save_content:
        if isinstance(browser_params.save_content, str):
            configured_types = set(browser_params.save_content.split(","))
            if not configured_types.issubset(ALL_RESOURCE_TYPES):
                diff = configured_types.difference(ALL_RESOURCE_TYPES)
                raise ConfigError(
                    "Unrecognized resource types provided ",
                    "in browser_params.save_content (%s)" % diff,
                )
```
this does not help our case to have `save_content` as a boolean as we can not assert `save_content` to be of a certain type(other than boolean) beforehand.

my solution would be to just add else statement to this if: `if isinstance(browser_params.save_content, str):` which raises an error but I am not sure if there is a case in whole codebase where `save_content` is set to `True` but is never set to a string

**Comment by vringar** [comment]
Okay, in this one case we will do type checking since the interface is really finicky to use.
Please emit an error if `save_content` is neither a string nor a boolean.

**Comment by ankushduacodes** [comment]
@vringar Could you please point me to documentation where I can find all the run configs that the platform checks for right now and what rules does it enforce as you have an example of `save_content` in this [comment](https://github.com/mozilla/OpenWPM/issues/495#issuecomment-732991428):
> how an entire crawl is configured so the combination of browser_params and manager_params. What validate_run_config is supposed to check is e.g.:

> * If you have content_saving enabled in one of the browser_params you should also have content_saving enabled in the manager_params (This is currently checked internally but shouldn't be)

**Comment by ankushduacodes** [comment]
`browser_params[i]["browser_id"] = self.data_aggregator.get_next_browser_id()`
and
`browser_params["profile_path"] = prof_folder`

@vringar, These are currently not part of default_browser_params
Do you suggest making them and others that I find part of dataclass defaults and set it to `None` or a `falsy`?

Also could you please point me to the documentation as requested [here](https://github.com/mozilla/OpenWPM/issues/495#issuecomment-734217405)

**Comment by vringar** [comment]
> @vringar Could you please point me to documentation (or point me to a file where run config rules are currently being enforced so I can have a look) where I can find all the run configs that the platform checks for right now and what rules does it enforce as you have an example of `save_content` in this [comment](https://github.com/mozilla/OpenWPM/issues/495#issuecomment-732991428):

Unfortunately there isn't much documentation on that. I think you found all of checks that we currently have in the codebase but I can take another look in the afternoon.

> `browser_params[i]["browser_id"] = self.data_aggregator.get_next_browser_id()`
> and
> `browser_params["profile_path"] = prof_folder`

I currently don't have a good solution for these yet. This is one of the reasons I'm now thinking it might be better to merge the PR as is and make changing the internal use a separate issue.

**Comment by ankushduacodes** [comment]
> > `browser_params[i]["browser_id"] = self.data_aggregator.get_next_browser_id()`
> > and
> > `browser_params["profile_path"] = prof_folder`
> 
> I currently don't have a good solution for these yet. This is one of the reasons I'm now thinking it might be better to merge the PR as is and make changing the internal use a separate issue.

From what I understand these parameters are always going to be overwritten by internal code. In my opinion it would not hurt to make them part of dataclasses. Even if users does set value for these while instanciating BrowserParams, it should not make any difference as explained above. And later we may be we can document that assigning values to these member variables is forbidden

**Comment by vringar** [comment]
I'm thinking maybe we should have `BrowserParamsInternal` and a `ManagerParamsInternal`classes that have all of these additional fields. They would inherit from `BrowserParams` and `ManagerParams` respectively and would be constructed after the validation.

**Comment by ankushduacodes** [comment]
Sounds good...🚀

**Comment by ankushduacodes** [comment]
@vringar I was looking for where the `ldb_address` gets set as it becomes part of the manager_params dict and I want to see what type it is of (probably tuple just like other address like `aggregator_address`).
Unfortunately, I could not find it, only usage I found was this:
https://github.com/mozilla/OpenWPM/blob/d0c34662500a40db38e0ecc5509dece9774aaa12/openwpm/deploy_browsers/deploy_firefox.py#L133-L136
